### ISSUE TITLE
v0.83 re-release?

### ISSUE THREAD

**Issue by notpeelz** [mentebinaria/readpe#211]
Why was the 0.83 tag recreated?
The new tag points to 62de24d2605547405a0c5811eda179232d13f728 (an empty commit?)

Previously, the v0.83 tag was on a version bump commit from 2023-11-28 with a hash of 7f3c13649959747ec57db172da355038c3521e1e.
Now on the 62de24d2605547405a0c5811eda179232d13f728 tree (pushed today), the version bump commit has a hash of c33268d25cb9573084cdc1ba107bcd776f6abe5d, suggesting the commit tree was rewritten.

So I dug into the commit history and found the most recent common ancestor, then looked at the differences between each commit after that.
It turns out [hh6d.golden.exe](https://github.com/mentebinaria/readpe/commit/9e4b950dbac0eedf28c3b6b3ff00db90aa613436#diff-6dd08c7626f70dd2fac1063f3f9b5bb96a2f699db31bbe0840b6a456d64c799a) (committed in  9e4b950dbac0eedf28c3b6b3ff00db90aa613436) was deleted, leaving it as an empty commit (now 62de24d2605547405a0c5811eda179232d13f728).

You can compare the two trees with this:
before: https://github.com/mentebinaria/readpe/compare/62de24d2605547405a0c5811eda179232d13f728...7f3c13649959747ec57db172da355038c3521e1e
after: https://github.com/mentebinaria/readpe/compare/7f3c13649959747ec57db172da355038c3521e1e...62de24d2605547405a0c5811eda179232d13f728

I thought maybe the file was malicious and you had to purge it from the repo, but [virustotal doesn't seem to think so](https://www.virustotal.com/gui/file/850a63631395d59b4dd83bd5d19eb9acc1066abbe8110c67d3f7396b8060c201/detection) (outside of a few false positives).
From what I could find on Google, `hh6d.golden.exe` appears to be a small test PE file downloaded [here](https://github.com/pts/pts-tinype). I verified the checksum and it matches the one from that repo (sha256: `850a63631395d59b4dd83bd5d19eb9acc1066abbe8110c67d3f7396b8060c201`).
Why go through all that trouble to get rid of an EXE file from the git history?

Rewriting a repo's history [can create a lot of issues](https://stackoverflow.com/a/1491022/1581233).
Changing existing git tags should also be avoided, as it will break many source-based packages. For example, on Arch Linux, a `PKGBUILD` published on the AUR would download `https://github.com/mentebinaria/readpe/archive/v0.83.tar.gz` and [validate it against a known-good checksum](https://wiki.archlinux.org/title/PKGBUILD#Integrity). Since the tag was re-created, it would now be impossible to download and build the software, as the checksum verification would fail.
Pinning versions (using checksums) is extremely important when it comes to supply chain security. Rewriting public branches breaks that.

-----

Here's what I suggest you do:

1. make sure your local repo is synced with the remote: `git fetch --all --tags`
2. checkout the master branch: `git checkout master`
3. restore the previous commit tree (locally): `git reset --hard 7f3c13649959747ec57db172da355038c3521e1e`
4. push the commit tree to the remote: `git push -f origin master`
5. delete the tag (locally) `git tag -d v0.83`
6. delete the tag from the remote `git push origin :refs/tags/v0.83`
7. re-create the tag to 7f3c13649959747ec57db172da355038c3521e1e: `git tag v0.83 7f3c13649959747ec57db172da355038c3521e1e`
8. push the tag: `git push origin v0.83`

If you really want to delete `hh6d.golden.exe`, you should revert that commit using `git revert`:
```
git revert 9e4b950dbac0eedf28c3b6b3ff00db90aa613436
git push origin master
```

Additionally, I suggest you switch to a 3-component version format (e.g [SemVer](https://semver.org/)), so that small changes like this can be released with a minor version bump (i.e. `0.83.0` -> `0.83.1`).

**Comment by notpeelz** [comment]
I just saw #210 
I should've read that before dissecting the git history, woops :P

**Comment by GoGoOtaku** [comment]
tl;dr (for people seeing just this issue) I had to purge the executable from the repository as it caused licensing.

I am aware that rewriting history is not something that should be done lightly. However it was in my opinion necessary given the issue.

I do not know how AUR handles package versioning. However I am a Gentoo user myself and in Gentoo it is possible to create a revision of a release. This is almost exclusively done for when the Gentoo build script was faulty but also when source location etc. changes.
I chose to not delete the empty commits as to have documentation why this happened.

If need be I could make a create a new tag/release.

**Comment by notpeelz** [comment]
> tl;dr (for people seeing just this issue) I had to purge the executable from the repository as it caused licensing.

Good point. I didn't realize the `pts-tinype` repo didn't specify a license. It could technically constitute a copyright violation, so it's better to err on the side of caution. I guess in this type of situation, you would need to contact the author to get permission to use their work. With that said, I understand your decision to rewrite the history, but maybe it would be better to simply delete the 0.93 release and re-release it as 0.94?

> I do not know how AUR handles package versioning. However I am a Gentoo user myself and in Gentoo it is possible to create a revision of a release. This is almost exclusively done for when the Gentoo build script was faulty but also when source location etc. changes. I chose to not delete the empty commits as to have documentation why this happened.

PKGBUILD has the `pkgrel` variable for handling package revisions. It's not so much that it can't be fixed, but that it breaks existing packages. It's not a big deal on rolling-release distros such as Arch (where packages aren't expected to be backward-compatible), but on distros like NixOS, [build reproducibility is very important](https://reproducible.nixos.org). A Nix derivation could pull packages using older, pinned, dependencies. If build artifacts aren't available on their cache servers, the derivation would be built locally (that's what happens with everything outside of [nixpkgs](https://github.com/NixOS/nixpkgs)). Nix dependency chains can get pretty complex, and random checksum changes like this are a *big* deal.

> If need be I could make a create a new tag/release.

:+1:

**Comment by GoGoOtaku** [comment]
I added a new release.

**Comment by notpeelz** [comment]
You might want to bump the version string to 0.84 (just like in #209)

**Comment by GoGoOtaku** [comment]
Thanks
